src/libostree/ostree-impl-system-generator.c

```diff
@@ -243,7 +250,7 @@ fstab_generator (const char *ostree_target, const bool is_aboot, const char *nor
                                "[Mount]\n"
                                "Where=%s\n"
                                "What=%s\n"
-                               "Options=bind\n",
+                               "Options=bind,slave,shared\n",
```

Comment:
I'm confused, aren't the states of being shared and slave mutually exclusive?

Reply:
Also unrelated to the above, we have a few places we mount `/var` unfortunately. There's some bits in `ostree-prepare-root.c`, especially if `INITRAMFS_MOUNT_VAR` is present.

I thought that at least some users relied on those and weren't actually running the fstab generator, FCOS included?

Reply:
Sorry, I should have been more explicit about this. The idea is to make the mount a slave of `/sysroot` and then create a separate shared group underneath that. From [mount_namespaces(7)](https://man7.org/linux/man-pages/man7/mount_namespaces.7.html#SHARED_SUBTREES):

> MS_SLAVE
> [mount(2)](https://man7.org/linux/man-pages/man2/mount.2.html) and [umount(2)](https://man7.org/linux/man-pages/man2/umount.2.html) events propagate into this mount
> from a (master) shared peer group.  [mount(2)](https://man7.org/linux/man-pages/man2/mount.2.html) and [umount(2)](https://man7.org/linux/man-pages/man2/umount.2.html)
> events under this mount do not propagate to any peer.
>
> Note that a mount can be the slave of another peer group
> while at the same time sharing [mount(2)](https://man7.org/linux/man-pages/man2/mount.2.html) and [umount(2)](https://man7.org/linux/man-pages/man2/umount.2.html)
> events with a peer group of which it is a member.  (More
> precisely, one peer group can be the slave of another peer
> group.)

What this means is that when you make a child mount of `/var`, it will be part of a shared group just like before. If you just make `/var` a slave to `/sysroot`, then a child mount of `/var` would become private. That's not what we want since then `/var/cache` would not propagate into an existing mount namespace.

This is show systemd sets up mounts when it creates a separate mount namespace. For example:

```
$ sudo nsenter -m -t $(systemctl show -P MainPID accounts-daemon.service) -- findmnt -l -o TARGET,PROPAGATION,OPT-FIELDS
TARGET                                                              PROPAGATION   OPT-FIELDS
/                                                                   shared,slave  shared:109 master:1
/run                                                                shared,slave  shared:125 master:13
/run/lock                                                           shared,slave  shared:126 master:14
/tmp                                                                shared,slave  shared:131 master:20
/efi                                                                shared,slave  shared:132 master:22
/dev                                                                shared        shared:133
/dev/pts                                                            shared,slave  shared:134 master:4
/dev/shm                                                            shared,slave  shared:135 master:3
/dev/mqueue                                                         shared,slave  shared:136 master:17
/dev/hugepages                                                      shared,slave  shared:137 master:16
/etc                                                                shared,slave  shared:138 master:1
/home                                                               shared,slave  shared:139 master:1
/proc                                                               shared        shared:140
/proc/sys/fs/binfmt_misc                                            shared,slave  shared:141 master:46
/proc/acpi                                                          shared        shared:142
/proc/asound                                                        shared        shared:143
/proc/bus                                                           shared        shared:144
/proc/fs                                                            shared        shared:145
/proc/irq                                                           shared        shared:146
/proc/kallsyms                                                      shared,slave  shared:147 master:13
/proc/kcore                                                         shared,slave  shared:148 master:13
/proc/mtrr                                                          shared        shared:149
/proc/scsi                                                          shared        shared:150
/proc/556/loginuid                                                  shared        shared:151
/proc/sys                                                           shared        shared:152
/proc/sys/fs/binfmt_misc                                            shared,slave  shared:153 master:46
/proc/sysrq-trigger                                                 shared        shared:154
/root                                                               shared,slave  shared:155 master:1
/run/credentials                                                    shared,slave  shared:127 master:13
/run/systemd/incoming                                               private,slave master:13
/run/systemd/seats                                                  shared,slave  shared:129 master:13
/run/user                                                           shared,slave  shared:130 master:13
/sys                                                                shared        shared:156
/sys/firmware/efi/efivars                                           shared,slave  shared:157 master:9
/sys/fs/bpf                                                         shared,slave  shared:158 master:10
/sys/fs/cgroup                                                      shared,slave  shared:159 master:7
/sys/fs/fuse/connections                                            shared,slave  shared:161 master:21
/sys/fs/pstore                                                      shared,slave  shared:162 master:8
/sys/kernel/config                                                  shared,slave  shared:163 master:11
/sys/kernel/debug                                                   shared,slave  shared:164 master:18
/sys/kernel/security                                                shared,slave  shared:165 master:6
/sys/kernel/tracing                                                 shared,slave  shared:166 master:19
/sys/fs/cgroup/system.slice/accounts-daemon.service/memory.pressure shared,slave  shared:160 master:7
/usr/lib/modules                                                    shared,slave  shared:167 master:13
/var/lib/AccountsService                                            shared,slave  shared:168 master:1
/var/log/lastlog                                                    shared,slave  shared:169 master:1
/var/mail                                                           shared,slave  shared:170 master:1
/efi                                                                shared,slave  shared:364 master:335
/run/user/1001                                                      shared,slave  shared:416 master:377
/run/user/1001                                                      shared,slave  shared:415 master:377
/run/user/1001/gvfs                                                 shared,slave  shared:593 master:587
/run/user/1001/gvfs                                                 shared,slave  shared:592 master:587
/run/user/1001/doc                                                  shared,slave  shared:826 master:794
/run/user/1001/doc                                                  shared,slave  shared:825 master:794
```

As for the `/var` mounting within prepare-root, I'm happy to fix that up. From my testing with cosa, it appears the generated `var.mount` is created and applied regardless of whether `/var` was mounted in the initramfs, but I agree it should DTRT out of the box.

Reply:
I added some more info in the comments and commit message about slave+shared propagation. I also updated the initramfs mounted `/var` case, too.